docs/Hooks.md

```diff
@@ -293,6 +294,32 @@ fastify.addHook('onRoute', (routeOptions) => {
   routeOptions.prefix
 })
 ```
+<a name="on-encapsulation"></a>
+**'onEncapsulation'**<br>
+Triggered when Fastify runs its internal logic that enables the encapsulation.<br/>
+This hook can be useful if you are developing a plugin that needs to use the encapsulation functionality of Fastify.
+```js
+fastify.decorate('data', [])
+
+fastify.register(async (instance, opts) => {
+  instance.data.push('hello')
+  console.log(instance.data) // ['hello']
+
+  instance.register(async (instance, opts) => {
+    instance.data.push('world')
+    console.log(instance.data) // ['hello', 'world']
+  })
+})
+
+fastify.register(async (instance, opts) => {
+  console.log(instance.data) // []
+})
+
+fastify.addHook('onEncapsulation', (instance) => {
+  instance.data = instance.data.slice()
```

Comment:
I am unclear what is happening in this example.

Reply:
I'm creating a new array from the old one, but without keeping the reference.
Basically is the same thing that we are doing inside our `override` function :)

Reply:
If you think that this example is too much obscure, I can refactor it with a `Object.assign` :)

Reply:
Maybe a dedicated file example with an `instance.route` could be more attractive?
Like "add on each context the `/schema` route that returns the `instance.getSchemas()`" (need to add prefix of course)
This to show one of the powers of this hook!

Reply:
I meant that I don't know what the whole example is showing, not just this one line.

Reply:
It shows that you can create more powerful encapsulation models for you decorators.